kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8167909?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8167909?buildTab=artifacts#/) on master @ [1d7bd69205c2197ccac33df9e2e6d4ff8c0fdbcf](https://github.com/cockroachdb/cockroach/commits/1d7bd69205c2197ccac33df9e2e6d4ff8c0fdbcf):


```
=== RUN   TestKVNemesisMultiNode
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode393864852
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:179: seed: 8655273352728583932
    kvnemesis_test.go:123: kvnemesis logging to /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2887003571
    kvnemesis.go:165: error applying x.AdminChangeReplicas(ctx, tk(12439996426192522190), [{REMOVE_VOTER n4,s4}]) // unable to find store 1 in range r55:/Table/100/"{1a3d4d94265ffad5"-d113bb11d3948624"} [(n4,s4):4, (n2,s2):6, (n3,s3):3, next=7, gen=19, sticky=9223372036.854775807,2147483647]: unable to find store 1 in range r55:/Table/100/"{1a3d4d94265ffad5"-d113bb11d3948624"} [(n4,s4):4, (n2,s2):6, (n3,s3):3, next=7, gen=19, sticky=9223372036.854775807,2147483647]
    kvnemesis.go:185: failures(verbose): /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2887003571/failures
        repro steps: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2887003571/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2887003571/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2887003571/kvs-scan.txt
    kvnemesis_test.go:206: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/577/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvnemesis/kvnemesis_test_/kvnemesis_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvnemesis/kvnemesis_test.go:206
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/577/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvnemesis/kvnemesis_test_/kvnemesis_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvnemesis/kvnemesis_test.go:159
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisMultiNode393864852
--- FAIL: TestKVNemesisMultiNode (19.35s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23038

--- a/pkg/kv/kvnemesis/validator.go
+++ b/pkg/kv/kvnemesis/validator.go
@@ -349,6 +349,35 @@ const (
 	bufferingBatchOrTxn
 )
 
+func transferLeaseResultIsIgnorable(res Result) bool {
+	err := errorFromResult(res)
+	if err == nil {
+		return false
+	}
+	return kvserver.IsLeaseTransferRejectedBecauseTargetMayNeedSnapshotError(err) ||
+		// Only VOTER (_FULL, _INCOMING, sometimes _OUTGOING) replicas can
+		// hold a range lease. Attempts to transfer to lease to any other
+		// replica type are rejected. See CheckCanReceiveLease.
+		resultIsErrorStr(res, `replica cannot hold lease`) ||
+		// Only replicas that are part of the range can be given
+		// the lease. This case is hit if a TransferLease op races
+		// with a ChangeReplicas op.
+		resultIsErrorStr(res, `replica not found in RangeDescriptor`) ||
+		// A lease transfer that races with a replica removal may find that
+		// the store it was targeting is no longer part of the range.
+		resultIsErrorStr(res, `unable to find store \d+ in range`) ||
+		// A lease transfer is not permitted while a range merge is in its
+		// critical phase.
+		resultIsErrorStr(res, `cannot transfer lease while merge in progress`) ||
+		// If the existing leaseholder has not yet heard about the transfer
+		// target's liveness record through gossip, it will return an error.
+		exceptLivenessCacheMiss(errorFromResult(res)) ||
+		// Same as above, but matches cases where ErrRecordCacheMiss is
+		// passed through a LeaseRejectedError. This is necessary until
+		// LeaseRejectedErrors works with errors.Cause.
+		resultIsErrorStr(res, liveness.ErrRecordCacheMiss.Error())
+}
+
 // processOp turns the result of an operation into its observations (which are
 // later checked against the MVCC history). The boolean parameter indicates
 // whether the operation is its own atomic unit or whether it's happening as
@@ -700,39 +729,15 @@ func (v *validator) processOp(op Operation) {
 			ignore = kvserver.IsRetriableReplicationChangeError(err) ||
 				kvserver.IsIllegalReplicationChangeError(err) ||
 				kvserver.IsReplicationChangeInProgressError(err) ||
-				errors.Is(err, roachpb.ErrReplicaCannotHoldLease)
+				errors.Is(err, roachpb.ErrReplicaCannotHoldLease) ||
+				transferLeaseResultIsIgnorable(t.Result) // replication changes can transfer leases
 		}
 		if !ignore {
 			v.failIfError(op, t.Result) // fail on all other errors
 		}
 	case *TransferLeaseOperation:
 		execTimestampStrictlyOptional = true
-		var ignore bool
-		if err := errorFromResult(t.Result); err != nil {
-			ignore = kvserver.IsLeaseTransferRejectedBecauseTargetMayNeedSnapshotError(err) ||
-				// Only VOTER (_FULL, _INCOMING, sometimes _OUTGOING) replicas can
-				// hold a range lease. Attempts to transfer to lease to any other
-				// replica type are rejected. See CheckCanReceiveLease.
-				resultIsErrorStr(t.Result, `replica cannot hold lease`) ||
-				// Only replicas that are part of the range can be given
-				// the lease. This case is hit if a TransferLease op races
-				// with a ChangeReplicas op.
-				resultIsErrorStr(t.Result, `replica not found in RangeDescriptor`) ||
-				// A lease transfer that races with a replica removal may find that
-				// the store it was targeting is no longer part of the range.
-				resultIsErrorStr(t.Result, `unable to find store \d+ in range`) ||
-				// A lease transfer is not permitted while a range merge is in its
-				// critical phase.
-				resultIsErrorStr(t.Result, `cannot transfer lease while merge in progress`) ||
-				// If the existing leaseholder has not yet heard about the transfer
-				// target's liveness record through gossip, it will return an error.
-				exceptLivenessCacheMiss(errorFromResult(t.Result)) ||
-				// Same as above, but matches cases where ErrRecordCacheMiss is
-				// passed through a LeaseRejectedError. This is necessary until
-				// LeaseRejectedErrors works with errors.Cause.
-				resultIsErrorStr(t.Result, liveness.ErrRecordCacheMiss.Error())
-		}
-		if !ignore {
+		if !transferLeaseResultIsIgnorable(t.Result) {
 			v.failIfError(op, t.Result) // fail on all other errors
 		}
 	case *ChangeZoneOperation:

